flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/connector/kafka/sink/FlinkKafkaInternalProducer.java

```diff
@@ -302,7 +302,7 @@ public void resumeTransaction(long producerId, short epoch) {
             transitionTransactionManagerStateTo(transactionManager, "READY");
 
             transitionTransactionManagerStateTo(transactionManager, "IN_TRANSACTION");
-            setField(transactionManager, "transactionStarted", true);
+            setField(transactionManager, "transactionStarted", false);
```

Comment:
I understand the reasoning to change this, but I'm not sure if it's the right thing to do.

@lightzhao can you clarify what this flag indicates? If it is `false`, does it tell the producer client to not send an explicit request to the Kafka transaction coordinator if we do `producer.commitTransaction()`? Then wouldn't things be incorrect in the case that there actually _is_ data in the transaction?

IIUC, this flag is used internally by the producer client because transactions are actually not considered started on the server side until the first `producer.send()` call. So, it uses this flag to decide whether or not to really send the `EndTxnRequest` request on a `producer.abortTransaction()` or `producer.commitTransaction()` API call.

Reply:
Please see my latest comment in the JIRA ticket.

I think we need to additionally persist this `transactionStarted ` flag in checkpoints as part of the `KafkaCommittable` metadata. And then set this flag appropriately to what is stored in the checkpoint.

By always setting it to `false`, transactions that actually have data written to them wouldn't be properly committed anymore.

Reply:
Ok, thanks for your reply, I have seen the reply from JIRA.
transactionStarted indicates that the partition was successfully added to the transaction or
Successfully added the consumer group's partitions to the transaction.
I tracked the status of transactionStarted of the kafka client. The transaction is always false from opening to sending completion until closing. I also tested that the sending is correct when there is data.

Reply:
But according to this line in the TransactionManager: https://github.com/a0x8o/kafka/blob/master/clients/src/main/java/org/apache/kafka/clients/producer/internals/TransactionManager.java#L789

if `transactionStarted` is false, then an `EndTxnRequest` will NOT be sent to Kafka - if there is data in the transaction, how is properly committed then?

Just to clarify: how are you verifying that data is sent properly? Are you using READ_COMMITTED mode on the consumer side?

Reply:
Ok, on the consumer side, I will use the READ_COMMITTED mode to verify.

Reply:
You are right, it is true that data cannot be consumed in READ_COMMITTED mode, and it is indeed impossible to directly change it to 'false'.